data/kernels/rgbcurve.cl

```diff
@@ -0,0 +1,90 @@
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "color_conversion.cl"
+
+kernel void
+rgbcurve(read_only image2d_t in, write_only image2d_t out, const int width, const int height,
+           read_only image2d_t table_r, read_only image2d_t table_g, read_only image2d_t table_b,
+           global float *coeffs_r, global float *coeffs_g, global float *coeffs_b,
+           read_only image2d_t lut_pow4, read_only image2d_t lut_pow5,
+           const int autoscale, const int preserve_colors,
+           global const dt_colorspaces_iccprofile_info_cl_t *profile_info, read_only image2d_t lut)
+{
+  const int x = get_global_id(0);
+  const int y = get_global_id(1);
+
+  if(x >= width || y >= height) return;
+
+  float4 pixel = read_imagef(in, sampleri, (int2)(x, y));
+
+  if(autoscale == 1) // DT_S_SCALE_MANUAL_RGB
+  {
+    pixel.x = lookup_unbounded(table_r, pixel.x, coeffs_r);
+    pixel.y = lookup_unbounded(table_g, pixel.y, coeffs_g);
+    pixel.z = lookup_unbounded(table_b, pixel.z, coeffs_b);
+  }
+  else if(autoscale == 0) // DT_S_SCALE_AUTOMATIC_RGB
+  {
+    if(preserve_colors == 1) // DT_RGBCURVE_PRESERVE_POWERNORM
+    {
+      const float R = (pixel.x > 0.f) ? pixel.x: 0.f;
+      const float G = (pixel.y > 0.f) ? pixel.y: 0.f;
+      const float B = (pixel.z > 0.f) ? pixel.z: 0.f;
+      
+      float ratio = 1.f;
+      const float div = get_lut_pow4(1.22f * R, lut_pow4) + get_lut_pow4(1.20f * G, lut_pow4) + get_lut_pow4(0.58f * B, lut_pow4);
+      if(div != 0.f)
+      {
+        const float weighted = 0.83743219f * ((get_lut_pow5(1.22f * R, lut_pow5) + get_lut_pow5(1.20f * G, lut_pow5) + 
+                                                get_lut_pow5(0.58f * B, lut_pow5)) / div);
+        if(weighted != 0.f)
+        {
+          const float curve_weighted = lookup_unbounded(table_r, weighted, coeffs_r);
+          ratio = curve_weighted / weighted;
+        }
+      }
+      
+      pixel.x = ratio * pixel.x;
+      pixel.y = ratio * pixel.y;
+      pixel.z = ratio * pixel.z;
+    }
+    else if(preserve_colors == 2) // DT_RGBCURVE_PRESERVE_LUMINANCE
+    {
+      float ratio = 1.f;
+      const float lum = get_rgb_matrix_luminance(pixel, profile_info, lut);
+      if(lum > 0.f)
+      {
+        const float curve_lum = lookup_unbounded(table_r, lum, coeffs_r);
+        ratio = curve_lum / lum;
+      }
+      
+      pixel.x = ratio * pixel.x;
+      pixel.y = ratio * pixel.y;
+      pixel.z = ratio * pixel.z;
+    }
+    else
+    {
+      pixel.x = lookup_unbounded(table_r, pixel.x, coeffs_r);
```

Comment:
looks suspicious that table_r and coeffs_r is used for the x, y & z values.

Reply:
This is the linked mode, all rgb use one curve.